cluster/src/main/scala/org/apache/spark/util/DynamicURLClassLoader.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2016 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package org.apache.spark.util
+
+import java.io.File
+import java.net.{URL, URLClassLoader}
+import java.util
+
+import scala.collection.JavaConverters._
+import scala.language.existentials
+
+import com.pivotal.gemfirexd.internal.engine.Misc
+import sun.misc.CompoundEnumeration
+import org.apache.spark.Logging
+private[spark] class DynamicURLClassLoader(urls: Array[URL],
+    parentClassLoader: ClassLoader, parentFirst: Boolean, overwriteFiles: Boolean = false)
+    extends MutableURLClassLoader(Array[URL](), parentClassLoader) with  Logging {
+  final private val JOB_SERVER_TEXT = "_SNAPPY_JOB_SERVER_JAR_"
+
+  private val urlList = new util.TreeMap[String, URLClassLoader]()
+
+  // this is executor class Path which is not going to change
+  private val parent = {
+    if (parentFirst) new MutableURLClassLoader(urls, parentClassLoader)
+    else new ChildFirstURLClassLoader(urls, parentClassLoader)
+  }
+
+  override def addURL(url: URL): Unit = {
+    // workaround to identify the files loaded by the job server as
+    // it always creates a new file with timestamp.
+    val key = if (url.getFile.lastIndexOf(JOB_SERVER_TEXT) > 0) {
+      url.getFile.substring(0, url.getFile.lastIndexOf(JOB_SERVER_TEXT))
+    } else url.getFile
+
+    val loader = getClassLoader(url)
+    def oldURL = if (urlList.containsKey(key)) urlList.get(key).getURLs.head.getPath else ""
+
+    if (!oldURL.isEmpty && !oldURL.equals(url.getPath)) {
+      // try to remove the old file. it is  okay if not successful
+      try {
+        logInfo(" Deleting the Loader with key:" + key + " with File: " + oldURL )
+        new File(oldURL).delete()
+      } catch {
+        case _: Throwable => {
+          // do nothing.
+        }
+      }
+    }
+
+    logInfo(" Adding new Loader with key:" + key + " with File: " + url.getPath )
+    urlList.put(key, loader)
+  }
+
+  private def getClassLoader(url: URL): MutableURLClassLoader = {
+    if (parentFirst) new MutableURLClassLoader(Array(url), parent)
+    else new ChildFirstURLClassLoader(Array(url), parent)
+  }
+
```

Comment:
Why double wrapping parent? Don't understand this additional URL handling here or above.